python/ray/data/tests/test_resource_manager.py

```diff
@@ -138,3 +145,145 @@ def test_calculating_usage(self):
         assert resource_manager.get_downstream_object_store_memory(o1) == 1700
         assert resource_manager.get_downstream_object_store_memory(o2) == 1700
         assert resource_manager.get_downstream_object_store_memory(o3) == 1000
+
+
+class TestReservationOpResourceLimiter:
+    """Tests for ReservationOpResourceLimiter."""
+
+    def test_basic(self, restore_data_context):
+        DataContext.get_current().op_resource_reservation_enabled = True
+
+        o1 = InputDataBuffer([])
+        o2 = MapOperator.create(MagicMock(), o1)
+        o3 = MapOperator.create(MagicMock(), o2)
+        o4 = LimitOperator(1, o3)
+
+        op_usages = {op: ExecutionResources.zero() for op in [o1, o2, o3, o4]}
+
+        topo, _ = build_streaming_topology(o4, ExecutionOptions())
+
+        global_limits = ExecutionResources.zero()
+
+        def mock_get_global_limits():
+            nonlocal global_limits
+            return global_limits
+
+        resource_manager = ResourceManager(topo, ExecutionOptions())
+        resource_manager.get_op_usage = MagicMock(side_effect=lambda op: op_usages[op])
```

Comment:
For my own understanding, when do you use `side_effect` rather than `return_value` in `MagicMock`?

Reply:
return_value is for returning a fixed value, while side_effect will replace the original function with a new function